src/merkle_tree.cairo

```diff
@@ -1,4 +1,39 @@
-use super::utils::double_sha256;
+use super::utils::{shr, double_sha256};
+
+#[derive(Copy, Drop)]
+pub struct Hash {
+    pub value: [u32; 8]
```

Comment:
This should not be public. It should have a constructor function. And Display and Debug implemented.

Reply:
`Display` should probably use `append_formatted_to_byte_array`

Reply:
You mean the whole struct not public or only the field ? I added it to the field just by anticipation if it needs to be accessed at the some point

Reply:
Do we need display + debug? Just one of them is not enough ?

Reply:
Not sure. You can do one with the other.

Reply:
from Cairo Book, printing section:

"Printing complex data types this way (via Display) might not be ideal as it requires additional steps to use the print! and println! macros. If you need to print complex data types, especially when debugging, use the Debug trait described below instead."

will go for Debug then. Moreover, Debug is already implemented in corelib for fixed size arrays, hence we can just derive Debug to make it work on our struct!

Reply:
@maciejka , @Jeanmichel7  rightly pointed out that we also need `PartialEq` to be implemented, especially for the case where we have an even number of leaves, because we need to do the following check of equality between 2 `Hash` struct :
```
  assert!(
            txids.at(len - 1) != txids.at(len - 2), "unexpected node duplication in merkle tree"
        );
 ```
 
 he will take care of the implementation in his PR that will follow this one

Reply:
> @maciejka , @Jeanmichel7 rightly pointed out that we also need `PartialEq` to be implemented, especially for the case where we have an even number of leaves, because we need to do the following check of equality between 2 `Hash` struct :
> 
> ```
>   assert!(
>             txids.at(len - 1) != txids.at(len - 2), "unexpected node duplication in merkle tree"
>         );
> ```
> 
> he will take care of the implementation in his PR that will follow this one

Teamwork!